staging/src/slime.io/slime/modules/meshregistry/pkg/source/zookeeper/watching.go

```diff
@@ -32,228 +34,445 @@ func (s *Source) ServiceNodeDelete(path string) {
+			return
+		default:
 		}
-		updateFunction(children, path)
-		for watchEvent := range e {
-			switch watchEvent.Type {
-			case zk.EventNodeDeleted:
-				if deleteFunction != nil {
-					deleteFunction(path)
-				}
-				s.watchPath.Remove(path)
-				scope.Infof("stop watch the zk path %q because the node was deleted", path)
-				return
-			case zk.EventResyncWatch:
-				scope.Infof("children under the zk path %q resync", path)
-				if children, _, err := s.Con.Children(path); err == nil {
-					updateFunction(children, path)
-				}
-			default:
-				// ignore other events as the channel will be closed unless the watch events are EventResyncWatch.
-				scope.Infof("children under the zk path %q maybe changed, rewatch and update it", path)
-			}
+
+		providers, _, eventCh, err := ew.conn.Load().(*zk.Conn).ChildrenW(ew.servicePath + providerPathSuffix)
+		if err != nil {
+			scope.Warnf("endpointWatcher %q's providerOnlyWatching watch provider failed: %v", ew.servicePath, err)
+			ew.once.Do(ew.initCallback)
+			continue
 		}
+		// ignore errors when getting consumer fails in provider watcher
+		consumers, _, _ := ew.conn.Load().(*zk.Conn).Children(ew.servicePath + consumerPathSuffix)
+		ew.handler(providers, consumers)
+		ew.once.Do(ew.initCallback)
+		<-eventCh
 	}
 }
 
-func (s *Source) doWatchInstance(path string, updateFunction func([]string, []string, string), deleteFunction func(string)) {
-	if s.watchPath.Has(path) {
-		return
+func (ew *EndpointWatcher) simpleWatch(path string, ch chan struct{}) {
+	b := backoff.Backoff{
+		Min: 500 * time.Millisecond,
+		Max: time.Minute,
 	}
-	scope.Infof("try to watch instance under the zk path: %q ", path)
-	initWatch := true
 	for {
-		pathChildren, _, pathEvent, err := s.Con.ChildrenW(path)
+		select {
+		case <-ew.exit:
+			return
+		default:
+		}
+		_, _, eventCh, err := ew.conn.Load().(*zk.Conn).ChildrenW(path)
 		if err != nil {
-			if shouldStopWatch(err) {
-				if !initWatch {
-					if deleteFunction != nil {
-						deleteFunction(path)
-					}
-					s.watchPath.Remove(path)
-				}
-				scope.Errorf("stop watch instance under the zk path %q with err: %v", path, err)
-				return
-			}
-			scope.Warnf("watch instance under the zk path %q failed: %s", path, err)
-			time.Sleep(time.Second)
+			time.Sleep(b.Duration())
+			continue
 		}
-		if initWatch {
-			s.watchPath.Set(path, nil)
-			scope.Infof("watching instance under the zk path: %q", path)
-			initWatch = false
+		b.Reset()
+		<-eventCh
+		select {
+		case ch <- struct{}{}:
+		default:
 		}
-		provider := false
-		providersPath := path[0:len(path)-10] + "/" + ProviderNode
-		consumersPath := path
-		if strings.HasSuffix(path, ProviderNode) {
-			provider = true
-			providersPath = path
-			consumersPath = path[0:len(path)-10] + "/" + ConsumerNode
+	}
+}
+
+func (ew *EndpointWatcher) watchBoth(ctx context.Context) {
+	defer close(ew.exit)
+
+	eventCh := make(chan struct{}, 1)
+	go ew.simpleWatch(ew.servicePath+providerPathSuffix, eventCh)
```

Comment:
Updating both providers and consumers when any of them changes unnecessarily magnifies the overhead

Reply:
so if only the consumer changes, we don't need to update?

Reply:
reuse the return of watch